api/src/services/data-ingestions/create-service.ts

```diff
@@ -0,0 +1,142 @@
+import Papa from "papaparse";
+import { DateTime } from "luxon";
+import { isNil } from "lodash";
+
+import { db } from "@/data/db-client";
+import {
+  DataIngestion,
+  DataIngestionMapping,
+  DataIngestionSource,
+  DataIngestionSourceNames,
+} from "@/data/models";
+import BaseService from "@/services/base-service";
+
+export class CreateService extends BaseService {
+  constructor(
+    private csvBuffer: Buffer,
+    private source_id: number,
+    private user_id: number
+  ) {
+    super();
+  }
+
+  async perform(): Promise<void> {
+    const csvText = this.csvBuffer.toString("utf-8");
+
+    const source = await this.getSourceOrThrow(this.source_id);
+    const rows = this.parseAndValidateCsv(source, csvText);
+
+    await this.clearDataIngestions(source, rows);
+    const mappings = await db("data_ingestion_mappings").where({ source_id: this.source_id });
+    const transformed = rows.map((row) => this.transformRow(row, mappings, source, this.user_id));
+    await db.transaction((trx) => trx.batchInsert("data_ingestions", transformed, 500));
+  }
+
+  async clearDataIngestions(
+    source: DataIngestionSource,
+    rows: Record<string, string>[]
+  ): Promise<void> {
+    const identifiers = rows
+      .map((row) => row[source.identifier_column_name])
+      .filter((id): id is string => Boolean(id));
+
+    if (identifiers.length === 0) {
+      return;
+    }
+
+    await db("data_ingestions")
+      .where({ source_id: source.id })
+      .whereIn("identifier", identifiers)
+      .delete();
+  }
+
+  private async getSourceOrThrow(source_id: number): Promise<DataIngestionSource> {
+    const source = await db("data_ingestion_sources").where({ id: source_id }).first();
+    if (isNil(source)) throw new Error(`Unknown source ID: ${source_id}`);
+    return source;
+  }
+
+  private parseAndValidateCsv(
+    source: DataIngestionSource,
+    csvText: string
+  ): Record<string, string>[] {
+    const lines = csvText.split(/\r?\n/);
+    const headerIndex = lines.findIndex((l) => l.includes(source.identifier_column_name));
+    if (headerIndex < 0) {
+      throw new Error(`Header row with '${source.identifier_column_name}' not found`);
+    }
```

Comment:
Can you make this more informative? e.g. `Invalid data format for Data Source "${data source}". Header row with ...`